### PR TITLE
WIP: Gapless playback

### PR DESCRIPTION
**Pull request by ashthespy**
Work in progress PR with the final goal of providing gapless playback.

### Todo
- [x] Implement a channel b/w `player` and `spirc` thread for prefetch 
- [ ] Prefetch next track when current one is fully downloaded (instead of time based approach now)
- [ ] Implement `duration` with other decoders as well if we are moving away from the time approach 
- [ ] Gapless playback 

#### Distinguish b/w playback state and loading state:
 Currently when the decoder is done with the current file, the player thread fires off an `EndOfTrack` to `spirc` - signalling the loading of the next track. For gapless, we thus need to implement a buffer (or is it a queue), and distinguish b/w fetching/decrypting/decoding states and actual playback state. 
Thoughts/ideas for this?

### CODE REVIEW

**Comment by sashahilton00** [comment]
While we're at it, we should probably add support for `kPlayStatusLoading` so that the client doesn't report the song as playing when it loads initially. This will require the player feeding info back to Spirc, in a manner similar to the preload trigger.

File: audio/src/lewton_decoder.rs [hunk 1]
```diff
@@ -24,6 +24,13 @@ where
         Ok(())
     }
 
+    pub fn postion(&self) -> Result<i64, VorbisError> {
```
Comment (sashahilton00) [inline audio/src/lewton_decoder.rs]: should this be `position` instead of `postion`?

Comment (ashthespy) [reply audio/src/lewton_decoder.rs]: It indeed should be - I was proofing another document today and noticed my editor has added this spelling to my auto suggest. 😯

Comment (sashahilton00) [reply audio/src/lewton_decoder.rs]: how helpful 🙃

File: playback/src/player.rs [hunk 2]
```diff
@@ -324,20 +377,27 @@ impl PlayerInternal {
...
+                let mut signal_prefetch = false;
                 let packet = if let PlayerState::Playing {
                     ref mut decoder,
                     normalisation_factor,
+                    duration,
                     ..
                 } = self.state
                 {
                     current_normalisation_factor = normalisation_factor;
+                    if i64::from(duration) - decoder.postion().unwrap() <= 2000 {
```
Comment (sashahilton00) [inline playback/src/player.rs]: same thing here.

**Comment by sashahilton00** [comment]
furthermore, we'll need to stop opening/closing the sink between songs, as there is a noticeable drop as sink opens/closes. Looks like we will need to implement a small buffer in player.rs after all.

**Comment by sashahilton00** [comment]
oops, accidentally left my 30s preload edit in there, sorry

**Comment by ashthespy** [comment]
Will fix up - I am testing a (very simple) buffer implementation - but need more inspiration. Got to read up on how this is normally done.

**Comment by sashahilton00** [comment]
Not sure how one would sync current state of track play with buffer. Presumably the start time of the next track is the time of the end of the previous track + the buffer size? Either way, we're going to have to stop releasing the sink between songs if we want gapless playback. Also, depending on how we design it, we may be able to extend the implementation to enable crossfade at a later date. I believe there was talk of switching to using rodio at somepoint, though that's a big task.

**Comment by jebos** [comment]
I was playing around with those changes. So one issue i found is that the code does not yet respect IF there is actually a 'next' Song. Its crashing if there is not, beside of that it seems to work okaish.

**Comment by jebos** [comment]
sorry for the delay: this is the BT I get
```
thread 'main' panicked at 'index out of bounds: the len is 1 but the index is 1', connect/src/spirc.rs:690:27
stack backtrace:
   0:   0x944ecf - std::sys::unix::backtrace::tracing::imp::unwind_backtrace::h63a67ac893ab31ec
                       at libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
   1:   0x933e13 - std::sys_common::backtrace::print::ha8f26509b9ff262c
                       at libstd/sys_common/backtrace.rs:71
                       at libstd/sys_common/backtrace.rs:59
   2:   0x93168b - std::panicking::default_hook::{{closure}}::hd7e08be4e57c3548
                       at libstd/panicking.rs:211
   3:   0x9312eb - std::panicking::default_hook::h216d888df769ef77
                       at libstd/panicking.rs:227
   4:   0x931db3 - std::panicking::rust_panic_with_hook::h7d39a7695d8da98f
                       at libstd/panicking.rs:477
   5:   0x931977 - std::panicking::continue_panic_fmt::ha287897123d4a956
                       at libstd/panicking.rs:391
   6:   0x93185f - rust_begin_unwind
                       at libstd/panicking.rs:326
   7:   0x995d73 - core::panicking::panic_fmt::h2155aa66b67fe83c
                       at libcore/panicking.rs:77
   8:   0x995d1f - core::panicking::panic_bounds_check::h083d97c982ea32c3
                       at libcore/panicking.rs:59
   9:   0x51203f - <librespot_connect::spirc::SpircTask as futures::future::Future>::poll::h2f783ae89b2b3c22
  10:   0x4e6743 - <librespot::Main as futures::future::Future>::poll::h3db72e6cc182213f
  11:   0x4c0c53 - futures::task_impl::std::set::h7c382029013f638f
  12:   0x4def7b - <futures::future::lazy::Lazy<F, R> as futures::future::Future>::poll::h339c16cdcdc8d2bf
  13:   0x4c0943 - futures::task_impl::std::set::h035c704851db453e
  14:   0x4c7a0f - <std::thread::local::LocalKey<T>>::with::h1726d6c9dbd47ab3
  15:   0x4ddfdf - <tokio::executor::current_thread::Entered<'a, P>>::block_on::h11c540a03761c4aa
  16:   0x4c7fcb - <std::thread::local::LocalKey<T>>::with::h5cfa747dd1304813
  17:   0x4c7b3b - <std::thread::local::LocalKey<T>>::with::h31d282700fc02d1b
  18:   0x4c7d77 - <std::thread::local::LocalKey<T>>::with::h41159fec93afe179
  19:   0x4c8c57 - <scoped_tls::ScopedKey<T>>::set::h9677cf0748583291
  20:   0x4da99b - tokio_core::reactor::Core::run::hf1b7213e939ae7d9
  21:   0x4e8887 - librespot::main::hffab8cbb1bfad5ce
  22:   0x4db287 - std::rt::lang_start::{{closure}}::h7734a4273cdf8ce1
  23:   0x9317d7 - std::panicking::try::do_call::he4497885895661a8
                       at libstd/rt.rs:59
                       at libstd/panicking.rs:310
  24:   0x952153 - __rust_maybe_catch_panic
                       at libpanic_unwind/lib.rs:103
  25:   0x93086b - std::panic::catch_unwind::hdd9654d0e8ee1176
                       at libstd/panicking.rs:289
                       at libstd/panic.rs:392
  26:   0x92d673 - std::rt::lang_start_internal::h4ec5eae1ccf41136
                       at libstd/rt.rs:58
  27:   0x4e9517 - main
  28: 0xb6d5d677 - __libc_start_main

```

**Comment by sashahilton00** [comment]
any progress on this out of interest?

**Comment by ashthespy** [comment]
> 
> 
> any progress on this out of interest?

Not much - I started rewriting the track loading and playback handling - but never got it more than a POC. Will get back to it soon hopefully - if someone feels like stepping in, be my guest!

**Comment by sashahilton00** [comment]
Not to worry, as just asking out of curiousity as much as anything, and also due to potential upcoming reimplementation of audio backends (move to Rodio), which might affect ongoing work on this PR. My current focus is to get rust protobuf and rodio PR's tested and merged, followed by publishing librespot to Cargo. This should then open up possibility of starting work on librespotd, whereby we can pull all the alternative audio backends into the daemon, then focus on issues which are actually relevant to the functioning of  the core library such as gapless/daily mixes/reconnection/etc.

**Comment by ashthespy** [comment]
This has slipped under my radar, will close it for now and pick it up when time permits.

**Pull request closed without merging**